rfcs/2020-08-27-3603-postgres-metrics.md

```diff
@@ -0,0 +1,154 @@
+# RFC 3603 - 2020-08-27 - Collecting metrics from PostgreSQL
+
+This RFC is to introduce a new metrics source to consume metrics from PostgreSQL database servers. The high level plan is to implement one source that collects metrics from PostgreSQL server instances.
+
+Background reading on PostgreSQL monitoring:
+
+- https://www.datadoghq.com/blog/postgresql-monitoring/
+
+## Scope
+
+This RFC will cover:
+
+- A new source for PostgreSQL server metrics.
+
+This RFC will not cover:
+
+- Other databases.
+
+## Motivation
+
+Users want to collect, transform, and forward metrics to better observe how their PostgreSQL databases are performing.
+
+## Internal Proposal
+
+Build a single source called `postgresql_metrics` (name to be confirmed) to collect PostgreSQL metrics.
+
+The recommended implementation is to use the Rust PostgreSQL client to connect the target database server by address specified in configuration.
+
+- https://docs.rs/postgres/0.17.5/postgres/index.html
+
+The source would then run the following queries:
+
+- `SELECT * FROM pg_stat_database`
+- `SELECT * FROM pg_stat_database_conflicts`
+- `SELECT * FROM pg_stat_bgwriter`
+
+And return these metrics by parsing the query results and converting them into metrics using the database name and column names.
+
+- `postgresql_up` -> Used as an uptime metric (0/1) ? - merits a broader discussion.
+- `pg_stat_database_blk_read_time` tagged with db, host, server, user (counter)
```

Comment:
Per https://github.com/timberio/vector/pull/3519 it seems like we want to move towards the [prometheus naming conventions](https://prometheus.io/docs/practices/naming/) for our metrics sources. For example, that would mean the counters here would typically be suffixed with `_total` and any metric with a unit would have the unit in the name. I think that would mean, for example, `pg_stat_database_blk_read_time` would be `pg_stat_database_blk_read_time_seconds_total` (I see the metric is in milliseconds, but prometheus naming conventions seem to prefer standardizing the units across metrics).

Oddly I see that the [semi-official prometheus postgresql exporter](https://github.com/wrouesnel/postgres_exporter) does not follow these conventions (I pasted the output I see below in the "prior art" section). I think we may still prefer to, though, to keep things consistent for vector users.

What is the `user` tag? Is it the connecting user? Or the user that owns the database being monitored?

I do note that the prometheus exporter tags with `datname` (database) _and_ `datid` (the database id). The latter I guess is nice as it never changes even if the database is renamed.

I echo Bruce's comment about the naming patterns too. It seems like they should be consistent? That is, everything prefixed with `postgresql` rather than `pg_stat` as the default namespace is `postgresql`.

Reply:
Yeah the prefix was a find/replace PEBCAK. I based these on the semi-official postgresql exporter but I can update them to the broader Prometheus naming standard. I'll leave a couple, like timestamps, that don't have endings specified.